compiler/modules/front/tests/testParser.bal

```diff
@@ -0,0 +1,91 @@
+import ballerina/test;
+import ballerina/io;
+
+import wso2/nballerina.err;
+
+@test:Config {
+    dataProvider: sourceFragments
+}
+function testParser(string k, string rule, string subject, string expected) returns err:Syntax|io:Error? {
+    Tokenizer tok = new (subject);
+    var tokErr = tok.advance();
+    test:assertTrue(!(tokErr is error), "tokenizer error for '" + subject + "'");
+    err:Syntax|Word[] parsed = reduceToWords(rule, tok);
+    if k.includes("F") {
+        if k.includes("V") {
+            test:assertTrue(parsed is err:Syntax, "test marked as failing but parsed for '" + subject + "'");
+            return;
+        }
+        if k.includes("E") {
+            test:assertTrue(parsed is Word[], "test marked as failing but correctly got an error for '" + subject + "'");
+            return;
+        }
+        err:unreached("kind must be FE or FV");
+    }
+    if k.includes("E") {
+        test:assertTrue(parsed is err:Syntax, "expected a syntax error for '" + subject);
+        return;
+    }
+    test:assertTrue(parsed is Word[], "syntax error for '" + subject + "'");
+    var actual = wordsToString(check parsed);
+    test:assertEquals(actual, expected, "wrong ast");
+    var last = tok.current();
+    test:assertEquals(last, (), "extra token '" + last.toString() + "' at the end for '" + subject + "'" );
+}
+
+function reduceToWords(string rule, Tokenizer tok) returns err:Syntax|Word[] {
+    Word[] w = [];
+    match rule {
+        "expr" => {
+            exprToWords(w, check parseExpr(tok));
+        }
+        "stmt" => {
+            stmtToWords(w, check parseStmt(tok));
+        }
+        _ => {
+            err:unreached("unknown production rule " + rule);
+        }
+    }
+    return w;
+}
+
+function sourceFragments() returns string[][] {
+    return [["E", "expr", "", ""],
+            // literals
+            ["V", "expr", "0", "0"],
+            ["V", "expr", "1", "1"],
+            ["FE", "expr", "01", ""],
+            ["V", "expr", "9223372036854775807", "9223372036854775807"],
+            ["E", "expr", "9223372036854775808", ""],
+            ["V", "expr", "true", "true"],
+            ["V", "expr", "false", "false"],
+            // unary op
+            ["FV", "expr", "+ 1", "+ 1"],
+            ["FV", "expr", "+1", "+ 1"],
```

Comment:
Unary `+` isn't in subset 1, so please comment these out for now. I think the normalized form of `+1` should be `+1` (with no space between the `+` and the `1`.